### PR TITLE
Some code optimization for gas amount

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Code refactoring: set() + executeOperation() -> executeOperation() + executeSingleSetOperation() + executeMultipleSetOperation()
- Move repeated gas amount logic to one place (executeSingleSetOperation()) 
- Generalize mergeGasAmounts() to mergeNumericJsObjects()
- Add 'test' to NATIVE_SERVICE_TYPES as a temporary solution
- Add more test cases
- Do some renaming:
  - totalGasAmount -> functionGasAmount
  - isServiceTx(), isAppTx() -> isServicePath(), isAppPath()
  - getSingleOpTotalGasAmount() -> getSingleOpServiceGasAmount()

Related issue: https://github.com/ainblockchain/ain-blockchain/issues/316, https://github.com/ainblockchain/ain-blockchain/issues/360

### CODE REVIEW

File: common/chain-util.js [hunk 1]
```diff
@@ -267,6 +267,16 @@ class ChainUtil {
     return true;
   }
 
+  static mergeNumericJsObjects(obj1, obj2) {
+    return _.mergeWith(obj1, obj2, (a, b) => {
+      if (!ChainUtil.isDict(a) && !ChainUtil.isDict(b)) {
+        if (ChainUtil.isNumber(a) && ChainUtil.isNumber(b)) {
```
Comment (liayoo) [inline common/chain-util.js:273]: question) did we decide to use isNumber() instead of numberOrZero()?

File: db/index.js [hunk 2]
```diff
@@ -862,12 +784,55 @@ class DB {
...
         return ChainUtil.returnTxResult(14, `Invalid operation type: ${op.type}`);
     }
+    if (!ChainUtil.isFailedTx(result)) {
+      const stateInfoAfter = this.getStateInfo('/');
+      const treeSizeDelta = stateInfoAfter[StateInfoProperties.TREE_SIZE] -
+          stateInfoBefore[StateInfoProperties.TREE_SIZE];
+      const gas = {
+        gas_amount: ChainUtil.getGasAmountObj(op.ref, 1, treeSizeDelta)
+      };
+      result.gas = ChainUtil.mergeNumericJsObjects(gas, result.gas);
```
Comment (liayoo) [inline db/index.js:797]: _.mergeWith(obj1, obj2) already modifies the obj1. Could we just do this instead? `ChainUtil.mergeNumericJsObjects(result.gas, gas);`

File: common/constants.js [hunk 3]
```diff
@@ -75,6 +75,7 @@ const NATIVE_SERVICE_TYPES = [
   'payments',
   'sharding',
   'staking',
+  'test',  // NOTE(seo): A temporary solution for tests.
```
Comment (liayoo) [inline common/constants.js:78]: 👍

File: common/chain-util.js (hunk 1 above)
Comment (platfowner) [reply common/chain-util.js]: Ah, just reverted and added some test cases (with null). Thanks!

File: db/index.js (hunk 2 above)
Comment (platfowner) [reply db/index.js]: It was for the case of undefined `result.gas`.
Just simplified the code by calling `mergeNumericJsObjects(result, { gas })`. Thanks!

**Comment by platfowner** [comment]
All comments addressed. PTAL @liayoo

**Comment by liayoo** [comment]
LGTM! Thanks for the nice refactoring! 🙌

**Review by liayoo — APPROVED** [review]

**Comment by platfowner** [comment]
Thanks for the review!

**Pull request merged**